src/main/java/com/lucidworks/spark/SchemaPreservingSolrRDD.java

```diff
@@ -0,0 +1,424 @@
+        if (name.substring(0, name.length() - 2).equals("__lwchilddocname")) {
+          finalName = field.getValue().toString();
+        } else {
+            fldr.add(new StructField(name.substring(0, name.length() - 2), getsqlDataType(field.getValue().toString()), true, Metadata.empty()));
+          }
+      }
+
+    }
+    StructField[] farr = new StructField[fldr.size()];
+    farr = fldr.toArray(farr);
+    StructType st2 = new StructType(farr);
+    if (finalName == null) {
+      finalName = "root";
+    }
+    return new StructField(finalName, st2, true,  Metadata.empty());
+  }
+
+  public static DataType getsqlDataType(String s) {
+    if (s.toLowerCase().equals("double")) {
+      return DataTypes.DoubleType;
+    }
+    if (s.toLowerCase().equals("byte")) {
+      return DataTypes.ByteType;
+    }
+    if (s.toLowerCase().equals("short")) {
+      return DataTypes.ShortType;
+    }
+    if (((s.toLowerCase().equals("int")) || (s.toLowerCase().equals("integer")))) {
+      return DataTypes.IntegerType;
+    }
+    if (s.toLowerCase().equals("long")) {
+      return DataTypes.LongType;
+    }
+    if (s.toLowerCase().equals("String")) {
+      return DataTypes.StringType;
+    }
+    if (s.toLowerCase().equals("boolean")) {
+      return DataTypes.BooleanType;
+    }
+    if (s.toLowerCase().equals("timestamp")) {
+      return DataTypes.TimestampType;
+    }
+    if (s.toLowerCase().equals("date")) {
+      return DataTypes.DateType;
+    }
+    if (s.toLowerCase().equals("vector")) {
+      return new VectorUDT();
+    }
+    if (s.toLowerCase().equals("matrix")) {
+      return new MatrixUDT();
+    }
+    if (s.contains(":") && s.split(":")[0].toLowerCase().equals("array")) {
+      return getArrayTypeRecurse(s,0);
+    }
+    return DataTypes.StringType;
+  }
+
+  public static DataType getArrayTypeRecurse(String s, int fromIdx) {
+    if (s.contains(":") && s.split(":")[1].toLowerCase().equals("array")) {
+      fromIdx = s.indexOf(":", fromIdx);
+      s = s.substring(fromIdx+1, s.length());
+      return DataTypes.createArrayType(getArrayTypeRecurse(s,fromIdx));
+    }
+    return DataTypes.createArrayType(getsqlDataType(s.split(":")[1]));
+  }
+
+  public JavaRDD<SolrDocument> queryShards(JavaSparkContext jsc, final SolrQuery origQuery, final String splitFieldName, final int splitsPerShard) throws SolrServerException {
+    origQuery.addFilterQuery("__lwcategory_s:data AND __lwroot_s:root");
+    return super.queryShards(jsc, origQuery, splitFieldName, splitsPerShard);
+  }
+
+  public JavaRDD<Row> toRows(final StructType schema, JavaRDD<SolrDocument> docs) {
+    JavaRDD<Row> rows = docs.map(new Function<SolrDocument, Row>() {
+      @Override
+      public Row call(SolrDocument solrDocument) throws Exception {
+        Row ret =  readData(solrDocument, getSolrClient(zkHost) ,  schema, collection);
```

Comment:
Since you're accessing an expensive object to allocate (SolrClient), you should wrap this in a foreach partition ... see foreachPartition in indexDocs in SolrSupport as an example.